### ISSUE TITLE
Issues with the current code

### ISSUE THREAD

**Issue by Polyhistor** [mitsouma/IOT_Anomalies_Detection_System_Based_on_Tensor_Decomposition#2]
Function uses global tensor_normal stats to normalize ALL tensors, creating severe data leakage.

```
def normalize_tensor_feature(tensor, feature_index):
    # MAJOR ISSUE: Using tensor_normal for ALL tensors!
    min_val = tensor_normal[:,:,:,feature_index].min() # par feature
    max_val = tensor_normal[:,:,:,feature_index].max() # par feature
    
    # This is applied to ALL tensors (train/test)
    normalized_slice = (feature_slice - min_val) / (max_val - min_val)
```

O(n) time_window lookup for every group
Inefficient nested loops
No vectorization

```
def create_graph_tensor(df, src_ip_to_idx, dst_ip_to_idx, window_size=300):
    # INEFFICIENT: O(n) lookup for each row
    for _, group in df.groupby(['time_window', 'Src IP', 'Dst IP']):
        t_idx = np.where(np.array(time_windows) == group['time_window'].iloc[0])[0][0]
        # Repeated dictionary lookups and array searches
```

Mathematical error in Khatri-Rao product order
Potential matrix dimension mismatch
No numerical stability checks

`def normal_space_projection(tensor_input, tensor_ref):
    # ISSUE: Inconsistent dimension handling
    kr_product = tl.tenalg.khatri_rao([D_ref, tl.tenalg.khatri_rao([C_ref, B_ref])])
    kr_pinv = torch.linalg.pinv(kr_product.T)  # Shape: (rank, I*J*K)
    
    for t in range(T):
        x_vec = tensor_input[t].reshape(1, -1)  # Shape: (1, I*J*K)
        A_proj[t] = x_vec @ kr_pinv  # Matrix mult with wrong dimensions`

Scaler is refit on each dataset, destroying normalization consistency.

```
# Training loop - REFITS scaler every time!
for df in df_train_list:
    X = scaled.fit_transform(df_cleaned.drop(columns=[...]))  # REFIT!
    
# Testing loop - REFITS again!  
for i, (model, df_scores_test) in enumerate(zip(model_list, df_test_list)):
    X_scaled = scaled.fit_transform(df_scores_test.drop(columns=[...]))  # REFIT!
```

All sigmoid activations → vanishing gradients
Dropout placed randomly
No batch normalization
Inconsistent layer sizing

`model = models.Sequential([
    layers.Dense(128, activation='sigmoid'),
    layers.Dense(80, activation='sigmoid'),
    layers.Dense(80, activation='sigmoid'),  # Duplicate layer
    layers.Dense(32, activation='sigmoid'),
    layers.Dense(32, activation='sigmoid'),  # Duplicate layer
    layers.Dropout(0.3),
    layers.Dense(64, activation='sigmoid'),  # Inconsistent ordering
    # ...
])`

No error handling for missing mappings
Fragile string-based feature matching
Silent failures possible

`def aggregation_scoring_Source(residuals, mappings, df_with_labels, window_size=300):
    # Incomplete mapping check
    if d in idx_to_dst_ip:
        dst_ip = idx_to_dst_ip[d]
        # Missing error handling for missing keys
    
    # String matching for features - fragile
    if 'count' in feature_name:
        aggregation[key]['count'] += score
    elif 'bytes' in feature_name:
        # What if feature_name doesn't match any pattern?`

The code below is really inefficient in memory consumption 

```
# Creates multiple large tensor copies
tensor_cpu = torch.zeros((len(time_windows), len(src_ip_to_idx), len(dst_ip_to_idx), 4))
tensor = tensor_cpu.to(device)  # Another copy!

# Multiple rank loops create huge memory overhead
for rank in range(1,20):
    factors_normal = parafac(tensor_normal, rank=rank, ...)
    factors_normal_list.append(factors_normal)  # Storing ALL decompositions
```

The code needs significant improvement.

**Comment by Polyhistor** [comment]
Thanks for the update @mitsouma , however the following still needs addressing 

 1. Memory Consumption Problems 

While this is partially addressed, you still have these following issues

Direct device tensor creation (no unnecessary CPU copies)
No longer stores all decompositions simultaneously Remaining Issues: Still processes all ranks sequentially (1-16), could be optimized, Better but not optimal


2. Scaler Consistency Problem seems to be still an issue ?

Original Issue: Scaler refit on each dataset destroying consistency
Current Code: SAME PROBLEM PERSISTS:


```
# Training
s = StandardScaler()
X1 = s.fit_transform(df_train_cleaned.drop(...))  # Fit on train

# Testing - WRONG!
s1 = StandardScaler()  # New scaler instance!
X = s1.fit_transform(df_test_cleaned.drop(...))  # Refit on test data!
```

 This causes severe data leakage
